src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1007,6 +1006,91 @@ RemainingTestsForCase makeNext(ArrayBuilder<BoundDagTest> remainingTests)
             }
         }
 
+        /// <summary>
+        /// Determine what we can learn from one successful runtime type test about another planned
+        /// runtime type test for the purpose of building the decision tree.
+        /// We accomodate a special behavior of the runtime here, which does not match the language rules.
+        /// A value of type `int[]` is an "instanceof" (i.e. result of the `isinst` instruction) the type
+        /// `uing[]` and vice versa.  It is similarly so for every pair of same-sized numeric types, and
+        /// arrays of enums are considered to be their underlying type.  We need the dag construction to
+        /// recognize this runtime behavior, so we pretend that matching one of them gives no information
+        /// on whether the other will be matched.  That isn't quite correct (nothing reasonable we do
+        /// could be), but it comes closest to preserving the existing C#7 behavior without undersireable
+        /// side-effects, and permits the code-gen strategy to preserve the dynamic semantic equivalence
+        /// of a switch (on the one hand) and a series of if-then-else statements (on the other).
+        /// See, for example, https://github.com/dotnet/roslyn/issues/35661
+        /// </summary>
+        bool? ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(TypeSymbol expressionType, TypeSymbol patternType, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            bool? result = Binder.ExpressionOfTypeMatchesPatternType(_conversions, expressionType, patternType, ref useSiteDiagnostics, out Conversion conversion);
+            return (!conversion.Exists && isRuntimeSimilar(expressionType, patternType))
+                ? null // runtime and compile-time test behavior differ. Pretend we don't know what happens.
+                : result;
+
+            static bool isRuntimeSimilar(TypeSymbol expressionType, TypeSymbol patternType)
+            {
+                while (expressionType is ArrayTypeSymbol { ElementType: var e1, IsSZArray: var sz1, Rank: var r1 } &&
```

Comment:
any need to test cases where the expression array type has a different rank than the pattern array type? I'd expect that a conversion wouldn't exist, so we wouldn't get here, but I don't know the spec deeply enough to say.

Reply:
We would get here, but we would conclude they are not runtime similar.  But even if we did, since the runtime doesn't treat them as similar the code would work correctly either way.  Are you suggesting some test like
``` c#
    case int[] a1: ...
    case Eint[,] a2: ...
```
?